scripts/stats.py

```diff
@@ -0,0 +1,194 @@
+import pynvml
+import time
+import os
+import threading
+import psutil
+import util
+import jsonlfile
+from numbers import Number
+from pathlib import Path
+
+EVENTS_FNAME = Path("events.jsonl")
+
+
+def gpu_in_use_by_this_process(gpu_handle):
+    if not psutil:
+        return False
+
+    base_process = psutil.Process().parent() or psutil.Process()
+
+    our_processes = base_process.children(recursive=True)
+    our_processes.append(base_process)
+
+    our_pids = set([process.pid for process in our_processes])
+
+    compute_pids = set(
+        [
+            process.pid
+            for process in pynvml.nvmlDeviceGetComputeRunningProcesses(gpu_handle)
+        ]
+    )
+    graphics_pids = set(
+        [
+            process.pid
+            for process in pynvml.nvmlDeviceGetGraphicsRunningProcesses(gpu_handle)
+        ]
+    )
+
+    pids_using_device = compute_pids | graphics_pids
+
+    return len(pids_using_device & our_pids) > 0
+
+
+class SystemStats(object):
+    def __init__(self):
+        try:
+            pynvml.nvmlInit()
+            self.gpu_count = pynvml.nvmlDeviceGetCount()
+        except pynvml.NVMLError:
+            self.gpu_count = 0
+
+        if psutil:
+            net = psutil.net_io_counters()
+            self.network_init = {"sent": net.bytes_sent, "recv": net.bytes_recv}
+        else:
+            print(
+                "psutil not installed, only GPU stats will be reported. \n Install with pip install psutil"
+            )
+
+        self._thread = threading.Thread(target=self._thread_body, daemon=True)
+        self._pid = os.getpid()
+        self._id = util.generate_id()
+        self._start_time = time.time()
+        self.sampler = {}
+        self.samples = 0
+        self._shutdown = False
+        self._dir = Path(
+            f"output/run_{time.strftime('%Y%m%d', time.localtime())}_{int(self._start_time // 10000)}-{self._id}"
+        )
+        self.events = jsonlfile.JsonlEventsFile(
+            self._start_time, EVENTS_FNAME, self._dir
+        )
+
+    def start(self):
+        self._thread.start()
+
+    @property
+    def proc(self):
+        return psutil.Process(pid=self._pid)
+
+    @property
+    def sample_rate_seconds(self):
+        """Sample system stats every this many seconds, default to 2"""
+        return 2
+
+    @property
+    def samples_to_average(self):
+        """The number of samples to average before pushing, default to 10"""
+        return 3
```

Comment:
`property`s should probably never return literal constants, but rather return private data members that you still want some way of exposing to the public API. These should get changed. Aso, the docstring for `samples_to_average` isn't correct.